File: packages/engine/src/skills/baseSkill.ts
import { TenantRegistryService } from 'business-configs';
import { SpiConnectorFactory, type ThirdPartySpiClient } from 'tools';
import type { AgentSkill, SkillExecutionContext, SkillExecutionResult, SkillMetadata, TenantSkillConfig } from 'types';

export abstract class BaseSkill implements AgentSkill {
  public abstract metadata: SkillMetadata;

  /**
   * 判断当前上下文是否由该 Skill 承接
   */
  public canHandle(context: SkillExecutionContext): boolean {
    const activeIntent = (context.slots?.activeIntent as string) || (context.extra?.intent as string) || '';
    return this.metadata.triggerIntents.includes(activeIntent);
  }

  /**
   * 获取租户针对当前 Skill 的个性化覆盖配置
   */
  public async getEffectiveConfig(tenantId: string): Promise<TenantSkillConfig | null> {
    const tenantConfig = await TenantRegistryService.getTenantConfig(tenantId);
    if (tenantConfig.skillsConfig && tenantConfig.skillsConfig[this.metadata.id]) {
      return tenantConfig.skillsConfig[this.metadata.id];
    }
    // 默认回落配置
    const isEnabled = !tenantConfig.enabledSkills || tenantConfig.enabledSkills.includes(this.metadata.id);
    return {
      skillId: this.metadata.id,
      enabled: isEnabled,
      approvalThresholdAmount: this.metadata.approvalThresholdAmount,
    };
  }

  /**
   * 获取当前生效的风控审批拦截金额阈值
   */
  public async getEffectiveApprovalThreshold(tenantId: string): Promise<number> {
    const config = await this.getEffectiveConfig(tenantId);
    if (config?.approvalThresholdAmount !== undefined) {
      return config.approvalThresholdAmount;
    }
    return this.metadata.approvalThresholdAmount ?? 50;
  }

  /**
   * 前置必要槽位或业务条件校验
   */
  public async validatePreconditions?(
    context: SkillExecutionContext,
  ): Promise<{ valid: boolean; missingPrompt?: string }>;

  /**
   * 核心业务 SOP 执行实现
   */
  public abstract execute(context: SkillExecutionContext): Promise<SkillExecutionResult>;

  /**
   * 快捷获取当前租户对应的 SPI 客户端 (动态适配本地 DB 或远程第三方 SPI)
   */
  protected async getSpiClient(tenantId: string): Promise<ThirdPartySpiClient> {
    const config = await TenantRegistryService.getTenantConfig(tenantId);
    return SpiConnectorFactory.getClient(config.spiConnector, tenantId);
  }
}

File: services/engine-py/src/engine_py/skills/base_skill.py
"""技能基类 — 租户配置覆盖 + 风控阈值 + SPI 客户端;进出为类型化契约。"""

from __future__ import annotations

from abc import ABC, abstractmethod

from ..tenant_config import get_tenant_config
from .contract import SkillContext, SkillResult
from .spi_client import LocalDbSpiAdapter


class BaseSkill(ABC):
    metadata: dict

    def can_handle(self, context: SkillContext) -> bool:
        active_intent = context.slots.get("activeIntent") or context.extra.get("intent") or ""
        return active_intent in self.metadata.get("triggerIntents", [])

    async def get_effective_config(self, tenant_id: str) -> dict | None:
        tenant_config = await get_tenant_config(tenant_id)
        skills_config = tenant_config.get("skillsConfig") or {}
        if self.metadata["id"] in skills_config:
            return skills_config[self.metadata["id"]]
        enabled_skills = tenant_config.get("enabledSkills")
        is_enabled = not enabled_skills or self.metadata["id"] in enabled_skills
        return {
            "skillId": self.metadata["id"],
            "enabled": is_enabled,
            "approvalThresholdAmount": self.metadata.get("approvalThresholdAmount"),
        }

    async def get_effective_approval_threshold(self, tenant_id: str) -> float:
        config = await self.get_effective_config(tenant_id)
        if config and config.get("approvalThresholdAmount") is not None:
            return float(config["approvalThresholdAmount"])
        return float(self.metadata.get("approvalThresholdAmount") or 50)

    @abstractmethod
    async def execute(self, context: SkillContext) -> SkillResult: ...

    async def get_spi_client(self, tenant_id: str) -> LocalDbSpiAdapter:
        """获取租户 SPI 客户端。TODO(Phase 1b):remote/mcp 模式适配器,当前统一本地适配器。"""
        return LocalDbSpiAdapter()
